aggregator/list.go

```diff
@@ -0,0 +1,342 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package aggregator
+
+import (
+	"bytes"
+	"container/list"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/m3db/m3metrics/metric"
+	"github.com/m3db/m3metrics/metric/aggregated"
+	"github.com/m3db/m3metrics/policy"
+	"github.com/m3db/m3metrics/protocol/msgpack"
+	"github.com/m3db/m3x/clock"
+	"github.com/m3db/m3x/log"
+)
+
+var (
+	errListClosed  = errors.New("metric list is closed")
+	errListsClosed = errors.New("metric lists are closed")
+)
+
+type encodeFn func(mp aggregated.ChunkedMetricWithPolicy) error
+
+// MetricList stores aggregated metrics at a given resolution
+// and flushes aggregations periodically
+// TODO(xichen): add metrics
+type MetricList struct {
+	sync.RWMutex
+
+	opts         Options
+	nowFn        clock.NowFn
+	log          xlog.Logger
+	timeLock     *sync.RWMutex
+	maxFlushSize int
+	flushFn      FlushFn
+	encoderPool  msgpack.BufferedEncoderPool
+
+	resolution    time.Duration
+	flushInterval time.Duration
+	aggregations  *list.List
+	timer         *time.Timer
+	idBuf         bytes.Buffer
+	encoder       msgpack.AggregatedEncoder
+	toCollect     []*list.Element
+	closed        bool
+	doneCh        chan struct{}
+	wgTick        sync.WaitGroup
+	encodeFn      encodeFn
+	aggMetricFn   aggMetricFn
+	waitForFn     waitForFn
+}
+
+func newMetricList(resolution time.Duration, opts Options) *MetricList {
+	// NB(xichen): by default the flush interval is the same as metric
+	// resolution, unless the resolution is smaller than the minimum flush
+	// interval, in which case we use the min flush interval to avoid excessing
+	// CPU overhead due to flushing
+	flushInterval := resolution
+	if minFlushInterval := opts.MinFlushInterval(); flushInterval < minFlushInterval {
+		flushInterval = minFlushInterval
+	}
+
+	encoderPool := opts.BufferedEncoderPool()
+	l := &MetricList{
+		opts:          opts,
+		nowFn:         opts.ClockOptions().NowFn(),
+		log:           opts.InstrumentOptions().Logger(),
+		timeLock:      opts.TimeLock(),
+		maxFlushSize:  opts.MaxFlushSize(),
+		flushFn:       opts.FlushFn(),
+		encoderPool:   encoderPool,
+		resolution:    resolution,
+		flushInterval: flushInterval,
+		aggregations:  list.New(),
+		timer:         time.NewTimer(0),
+		encoder:       msgpack.NewAggregatedEncoder(encoderPool.Get()),
+		doneCh:        make(chan struct{}),
+	}
+	l.encodeFn = l.encoder.EncodeChunkedMetricWithPolicy
+	l.aggMetricFn = l.processAggregatedMetric
+	l.waitForFn = time.After
+
+	// Start ticking
+	if l.flushInterval > 0 {
+		l.wgTick.Add(1)
+		go l.tick()
+	}
+
+	return l
+}
+
+// Close closes the list
+func (l *MetricList) Close() {
+	l.Lock()
+	if l.closed {
```

Comment:
Yeah basically I'm using a bool if there is already a `RWMutex` in the object to save the overhead of atomics, and a uint32 otherwise